queue/retry.go

```diff
@@ -0,0 +1,256 @@
+package queue
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+)
+
+type retryHandler struct {
+	// kim: TODO: should add a RetryableQueue interface with an extra PutRetry()
+	// method. That way, it's optional to implement the retryability features.
+	queue         amboy.RetryableQueue
+	opts          amboy.RetryHandlerOptions
+	pending       map[string]amboy.RetryableJob
+	started       bool
+	wg            sync.WaitGroup
+	mu            sync.RWMutex
+	cancelWorkers context.CancelFunc
+}
+
+func newRetryHandler(q amboy.RetryableQueue, opts amboy.RetryHandlerOptions) (amboy.RetryHandler, error) {
+	if q == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+	if err := opts.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid options")
+	}
+	return &retryHandler{
+		queue:   q,
+		opts:    opts,
+		pending: map[string]amboy.RetryableJob{},
+	}, nil
+}
+
+func (rh *retryHandler) Start(ctx context.Context) error {
+	rh.mu.Lock()
+	defer rh.mu.Unlock()
+
+	rh.started = true
+
+	workerCtx, workerCancel := context.WithCancel(ctx)
+	rh.cancelWorkers = workerCancel
+	for i := 0; i < rh.opts.NumWorkers; i++ {
+		rh.wg.Add(1)
+		go func() {
```

Comment:
Can we combine this go routine with `waitForJob`?

Reply:
I don't know what you mean. This is starting a goroutine to run waitForJob asynchronously and dealing with restarting it in the defer panic handler. I wouldn't put the goroutine inside waitForJob.